sqlc: user and organization

PaulKalho opened the pull request:
## Which issues does this pull request close
Closes #573, #574

## Important
- **AvatarUrl** made optional in GetOrganizationResponse
- **AvatarUrl** made optional in GetOrganizationsByUserResponse
- **AvatarUrl** made optional in GetOrganizationsForUserResponse
- **AvatarUrl** made optional in GetInvitationsByUserRespoonse

### Tests
Unfortunately I also had to edit some tests because I encountered an error which I didn't while writing the tests 

Affected: 
- GetInvitationsByOrganizations
- AddRemoveMember (bec. is using GetMembersByOrganization in test)
- GetMembersByOrganization 

=> All these Endpoints used the function `IsInOrganization` which was implemented wrong. It always returned true.

- RevokeInvitation

=> I'm not quite sure what gorm did there but I always got a wrong row in the query.

FoseFx commented on services/user-svc/repos/organization_repo.sql:
```diff
@@ -0,0 +1,122 @@
+	SELECT 1
+	FROM invitations
+	WHERE (invitations.email = @email AND invitations.organization_id = @organization_id)
+		AND state = ANY(ARRAY[@state::int[]])
```
```suggestion
		AND state = ANY(ARRAY[@states::int[]])
```

FoseFx commented on services/user-svc/repos/organization_repo.sql:
```diff
@@ -0,0 +1,122 @@
+SELECT EXISTS (
+	SELECT 1
+	FROM invitations
+	WHERE (invitations.email = @email AND invitations.organization_id = @organization_id)
```
nit: `invitations.` is not needed here

FoseFx commented on services/user-svc/repos/organization_repo.sql:
```diff
@@ -0,0 +1,122 @@
+FROM invitations
+	JOIN organizations ON invitations.organization_id = organizations.id
+WHERE email = @email
+	AND state = coalesce(sqlc.narg('state'), state);
```
an OR ... IS NULL might be faster here

Without knowing about what's going on internally, I believe this has to pull out the state of every row and compare it to itself, if state is NULL

FoseFx reviewed (changes requested):
Nice work! Some of the tests need to be looked into though.

PaulKalho commented:
@FoseFx thanks for the Review. 
I've cleaned up the sql statements and refactored the tests. Left the `not a member` tests inside, but added the missing ones.

FoseFx commented on services/user-svc/internal/organization/organization.go:
```diff
@@ -685,88 +701,66 @@ func (s ServiceServer) RevokeInvitation(ctx context.Context, req *pb.RevokeInvit
 	})
-
 	if err != nil {
 		return nil, err
```
I believe you have to rollback the transaction before we return. From what I read in the docs you could simply add a `defer tx.Rollback()` after creating the transaction, which will cause the rollback to be attempted before any return. If we commit before, the rollback won't happen. But please read up on that again :D

PaulKalho replied:
Yes you are 100% right, must have missed that.

FoseFx commented on services/user-svc/internal/user/user.go:
```diff
@@ -31,86 +34,93 @@ func NewServiceServer() *ServiceServer {
 		Name:      user.Name,
 		Nickname:  user.Nickname,
-		AvatarUrl: user.Avatar,
+		AvatarUrl: *user.AvatarUrl,
```
AvatarUrl is not `NOT NULL`.
THIS WILL CAUSE NIL-DEREFERENCES!

PaulKalho replied:
Hm true, thought because of the default value this shouldn't be a problem.
Forgot that it can theoretically be NULL if this was inserted intentionally, will fix that.

FoseFx reviewed (changes requested):
Mostly small things, but the transaction rollback and NIL-deref bugs *need* to be fixed

PaulKalho commented on services/user-svc/internal/organization/organization.go:
```diff
@@ -326,9 +344,9 @@ func (s ServiceServer) RemoveMember(ctx context.Context, req *pb.RemoveMemberReq
 }
 
 func (s ServiceServer) InviteMember(ctx context.Context, req *pb.InviteMemberRequest) (*pb.InviteMemberResponse, error) {
```
that's a good call, ty

MaxSchaefer commented:
@FoseFx

FoseFx commented on services/user-svc/repos/organization_repo.sql:
```diff
@@ -0,0 +1,139 @@
+		WHERE memberships.organization_id = @organization_id
+		  AND users.email = @email
+		LIMIT 1
+	) AS isInOrganitationByEmail,
```
```suggestion
	) AS is_in_organization_by_email,
```

FoseFx commented on services/user-svc/repos/organization_repo.sql:
```diff
@@ -0,0 +1,139 @@
+		SELECT 1
+		FROM organizations
+		WHERE (id = @organization_id)
+	) AS doesOrganizationExist,
```
```suggestion
	) AS does_organization_exist,
```

FoseFx commented on services/user-svc/repos/organization_repo.sql:
```diff
@@ -0,0 +1,139 @@
+		FROM invitations
+		WHERE (email = @email AND organization_id = @organization_id)
+		  AND state = ANY(ARRAY[@states::int[]])
+	) AS DoesInvitationExist;
```
```suggestion
	) AS does_invitation_exist;
```

FoseFx commented on services/user-svc/repos/organization_repo.sql:
```diff
@@ -0,0 +1,139 @@
+		SELECT 1
+		FROM invitations
+		WHERE (email = @email AND organization_id = @organization_id)
+		  AND state = ANY(ARRAY[@states::int[]])
```
what does the `ARRAY[..]` do / why is it needed?

PaulKalho replied:
You're right, we dont't need it at all. Wonder why I put it there originally...

FoseFx commented on services/user-svc/internal/organization/organization.go:
```diff
@@ -709,6 +692,10 @@ func CreateOrganizationAndAddUser(ctx context.Context, attr organization_repo.Or
 	}
 
+	defer func() {
+		_ = tx.Rollback(ctx)
```
```suggestion
		// TODO: log Rollback errors
		_ = tx.Rollback(ctx)
```

FoseFx reviewed (approved):
Nice! Only a few finishing touches, then you can merge!
